### PR TITLE
DTM density estimator

### PR DESCRIPTION
**Pull request by mglisse**
I think this is the last piece broken out from tomato.

Doc: https://1095-175032741-gh.circle-artifacts.com/0/tmp/sphinx/index.html

### CODE REVIEW

File: src/python/gudhi/point_cloud/dtm.py [hunk 1]
```diff
@@ -68,3 +69,110 @@ def transform(self, X):
...
+        We recommend using a small fixed value instead in those cases, even if it won't have the same nice
+        theoretical properties as the dimension.
+    """
+
+    def __init__(self, k=None, weights=None, q=None, dim=None, normalize=False, n_samples=None, **kwargs):
+        """
+        Args:
+            k (int): number of neighbors (possibly including the point itself). Optional if it can be guessed
+                from weights or metric="neighbors".
+            weights (numpy.array): weights of each of the k neighbors, optional. They are supposed to sum to 1.
```
[the remark this answers is not in the document]
Comment (mglisse) [reply src/python/gudhi/point_cloud/dtm.py:90]: I don't know. I didn't want to forbid it. A warning... maybe, I don't have a strong opinion there.

File: src/python/test/test_dtm.py [hunk 2]
```diff
@@ -66,3 +67,22 @@ def test_dtm_precomputed():
...
+
+
+def test_density():
+    distances = [[0, 1, 10], [2, 0, 30], [1, 3, 5]]
+    density = DTMDensity(k=2, metric="neighbors", dim=1).fit_transform(distances)
+    expected = numpy.array([2.0, 1.0, 0.5])
+    assert density == pytest.approx(expected)
+    distances = [[0, 1], [2, 0], [1, 3]]
+    density = DTMDensity(metric="neighbors", dim=1).fit_transform(distances)
+    expected = numpy.array([2.0, 1.0, 0.5])
```
[the remark this answers is not in the document]
Comment (mglisse) [reply src/python/test/test_dtm.py:87]: True. It helps make the tests independent, so modifying one has no effect on the other, but in this case I could indeed remove it.

File: src/python/test/test_dtm.py [hunk 3]
```diff
@@ -66,3 +67,22 @@ def test_dtm_precomputed():
...
+    assert density == pytest.approx(expected)
```
[the remark this answers is not in the document]
Comment (mglisse) [reply src/python/test/test_dtm.py:88]: I thought I had one. I'll add it.

File: src/python/gudhi/point_cloud/dtm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/gudhi/point_cloud/dtm.py:90]: Yes, maybe a warning. I let you decide.

File: src/python/test/test_dtm.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/test/test_dtm.py]: If you want to factorize some part of code and split your tests, you can always use [fixtures](https://docs.pytest.org/en/latest/fixture.html).

File: src/python/gudhi/point_cloud/dtm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mglisse) [reply src/python/gudhi/point_cloud/dtm.py:90]: There is also the possibility of renormalizing the weights inside the function. If the weights don't sum to 1, that just multiplies the result with a constant. Unless you specifically ask for `normalize=True`, the output is only defined up to a multiplicative function anyway, so it doesn't really look that bad if the weights don't sum to 1... Maybe I'll add a warning just when `normalize=True`, or renormalize in that case.

[the remark this answers is not in the document]
Comment (mglisse) [reply src/python/gudhi/point_cloud/dtm.py:90]: Actually, when `normalize=True`, any multiplicative factor on the weights gets cancelled out automatically. So I simply won't warn.

[the remark this answers is not in the document]
Comment (VincentRouvreau) [reply src/python/gudhi/point_cloud/dtm.py:90]: That's fine to me.
